### PR TITLE
feature ssh connection

### PR DESCRIPTION
**Pull request by Zaher1307**
Making two functions one for making the connection with the VM using the IP address and returning the connection object, and another function for sending commands to the VM.

### CODE REVIEW

**Review by IslamWalid — CHANGES_REQUESTED** [review]

File: src/services/ssh.js [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+      reject(err);
+    }).connect(config);
+  });
+}
+
+function runCommand(conn, command) {
+  return new Promise((resolve, reject) => {
+    conn.exec(command, (err, stream) => {
+      if (err) {
+        reject(err);
```
Comment (IslamWalid) [inline src/services/ssh.js:18]: you need to return in the `if-statement`

Comment (IslamWalid) [reply src/services/ssh.js:18]: you need to create an error handler module to handle the errors in the connecting process

Comment (Zaher1307) [reply src/services/ssh.js:18]: Ok, I fixed it

Comment (IslamWalid) [reply src/services/ssh.js:18]: It's a good style to add an extra space between the error handling `if` statement and the rest of the code.

File: src/services/ssh.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+      }
```
Comment (IslamWalid) [inline src/services/ssh.js:19]: It's a good style to add an empty line between the error handling if statement and the rest of the code.

Comment (Zaher1307) [reply src/services/ssh.js:19]: I fixed it

File: src/services/ssh.js
[the remark this answers is not in the document]
Comment (Zaher1307) [reply src/services/ssh.js:1]: Ok, I did it

**Review by meldisoukyy — APPROVED** [review]
Awesome!

File: src/services/ssh.js [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+  return new Promise((resolve, reject) => {
+    conn.exec(cmd, (err, stream) => {
+      if (err) {
+        return reject(err);
+      }
+
+      let result = '';
+      let error = '';
+      stream.on('close', (code, signal) => {
+        resolve(result || error);
```
Comment (meldisoukyy) [inline src/services/ssh.js:24]: I think it is better to distinguish between the result and the error

**Review by meldisoukyy — APPROVED** [review]

File: app.js [hunk 4]
```diff
@@ -0,0 +1,9 @@
+const express = require('express');
+const sshRoute = require('./src/routes/ssh');
```
Comment (meldisoukyy) [inline app.js:2]: you have to change the path of `sshRoute` to be `'./src/services/ssh'`,
and it's better to rename `sshRoute` to be `sshService` for example.

File: src/services/ssh.js (hunk 3 above)
Comment (meldisoukyy) [inline src/services/ssh.js:24]: I think it's better to distinguish between the result and the error.

**Review by meldisoukyy — CHANGES_REQUESTED** [review]

**Review by IslamWalid — CHANGES_REQUESTED** [review]

File: app.js (hunk 4 above)
Comment (IslamWalid) [inline app.js:2]: `ssh` won't be a route at all, it's just a service. There's no need to add it as a route in the app in the first place

**Review by IslamWalid — CHANGES_REQUESTED** [review]

File: src/services/ssh.js (hunk 3 above)
Comment (Zaher1307) [reply src/services/ssh.js:24]: do you have any suggestions?

File: src/services/ssh.js [hunk 5]
```diff
@@ -0,0 +1,34 @@
+const { Client } = require('ssh2');
+const conn = new Client();
+
+function connect(config) {
```
[the remark this answers is not in the document]
Comment (Zaher1307) [reply src/services/ssh.js:4]: I think there is not the right place to check for errors, for example, if I have a controller to connect it should check first for all credentials

File: app.js (hunk 4 above)
Comment (Zaher1307) [reply app.js:2]: yes it's a mistake I had a module to test the service then I forget to remove app.js and index.js

File: src/services/ssh.js [hunk 6]
```diff
@@ -0,0 +1,34 @@
+const { Client } = require('ssh2');
```
[the remark this answers is not in the document]
Comment (Zaher1307) [reply src/services/ssh.js:1]: ok it's my next sprint

File: src/services/ssh.js [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+      resolve(conn);
+    }).on('error', (err) => {
+      reject(err);
+    }).connect(config);
+  });
+}
+
+conn.runCmd = function runCmd(cmd) {
+  return new Promise((resolve, reject) => {
+    conn.exec(cmd, (err, stream) => {
```
[the remark this answers is not in the document]
Comment (Zaher1307) [reply src/services/ssh.js]: Done

File: src/services/ssh.js (hunk 3 above)
Comment (meldisoukyy) [reply src/services/ssh.js:24]: you may resolve an object like this {type: "result"| "error", content: result|error}

Comment (Zaher1307) [reply src/services/ssh.js]: nice, I made a similar solution

**Review by meldisoukyy — APPROVED** [review]

**Comment by Zaher1307** [comment]
I have found a wrapper package doing what I need.
the package called [`node-ssh`](https://www.npmjs.com/package/node-ssh)

**Pull request closed without merging**